File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()
        for index,value in enumerate(nums):
            if index > 0 and value == nums[index-1]:
                continue
            l,r=index+1,len(nums)-1
            while(l<r):
                total = value + nums[l] + nums[r]
                if (total>0):
                    r = r-1
                elif total<0:
                    l = l+1
                else:
                    result.append([value,nums[l],nums[r]])
                    l = l+1
                    while nums[l]==nums[l-1] and l<r:
                        l= l+1
        return result

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
        let result = [];
        nums.sort((a, b) => a - b);
        for (let index = 0; index < nums.length; index++) {
            if (index > 0 && nums[index] === nums[index - 1]) {
                continue;
            }
            let l = index + 1;
            let r = nums.length - 1;
            while (l < r) {
                let total = nums[index] + nums[l] + nums[r];
                if (total > 0) {
                    r--;
                } else if (total < 0) {
                    l++;
                } else {
                    result.push([nums[index], nums[l], nums[r]]);
                    l++;
                    while (nums[l] === nums[l - 1] && l < r) {
                        l++;
                    }
                }
            }
        }
        return result;
};